### PR TITLE
Add macro that faciliates "bazel run" of binary generated by rules_foreign_cc

### LINKED ISSUE #582: Can't `bazel run` things generated as `out_binaries`
I put together a `rules_foreign_cc` rule for building a binary which builds with configure+make, and wanted to be able to `bazel run` it.

This was tricky - a bunch of the things I wanted to do didn't work:

First I tried to just `bazel run` the target, but the target isn't labelled as `executable`. Rules can't be conditionally marked `executable` based on e.g. whether they have `out_binaries` (or exactly one `out_binary`). We could do something like "always run the first `out_binary` if there are any, and generate a dummy executable if not", but that seems kind of sketchy, and ideally the `run`-ability could be decided at analysis time.

Next I tried to copy the generated file in a `genrule` marked `executable=True`, but the generated file didn't have an associated label (for `//target:my` with an `out_binary` named `binary`, I couldn't write `//target:my/bin/binary`).

So I ended up writing a wrapper rule, `runnable_binary` which depends on the `rules_foreign_cc` target, looks up the file in the provider and does the copy, and is marked as `executable`.

I can upstream `runnable_binary` if folks would be interested, but I'd be interested in whether there's a better approach here...

### PR DESCRIPTION
**Pull request by jheaff1**
This resolves #582 

The macro also facilitates the running of dynamically linked executables, which could then be used as tools in dependent targets. This is useful when using LGPL licenced third party software that must be dynamically linked.

Note that the OpenSSL example has been changed to produce shared libraries and to produce the `openssl` binary. This tests that `runnable_binary` can successfully run a dynamically linked binary

### CODE REVIEW

**Comment by jheaff1** [comment]
@UebelAndre @jsharpe  Would either of you be able to give me a hand with the MacOS failures in CI? 

It seems now that OpenSSL shared libraries are built rather than static libraries, the curl test fails on MacOS only.

**Comment by jsharpe** [comment]
> @UebelAndre @jsharpe Would either of you be able to give me a hand with the MacOS failures in CI?
> 
> It seems now that OpenSSL shared libraries are built rather than static libraries, the curl test fails on MacOS only.

I haven't got time to spend much time on this right now but looks to be rpath related. I'm guessing the rpath in the binary isn't looking at the right path to find the shared library in the runfiles?

**Comment by jheaff1** [comment]
I think the MacOS issue is related to bazelbuild/bazel#10254

**Comment by jheaff1** [comment]
@jsharpe @UebelAndre As it seems that the use of shared libs on sandboxed macos is a bit broken, i have removed the failing tests from CI. I hope that's acceptable.

File: foreign_cc/private/runnable_binary_wrapper.sh [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+# Allow unbound variable here, in case LD_LIBRARY_PATH or similar is not already set
+set +u
+for dir in "${SHARED_LIBS_DIRS_ARRAY[@]}"; do
+    export ${LIB_PATH_VAR}="${!LIB_PATH_VAR}":"$dir"
+done
+set -u
+
+EXE=BIN
+$(rlocation "${EXE#external/}") $@
```
Comment (jsharpe) [inline foreign_cc/private/runnable_binary_wrapper.sh:48]: exec avoids creating another process and the quotes mean that spaces in any arguments get handled correctly.
```suggestion
exec $(rlocation "${EXE#external/}") "$@"
```

**Review by jsharpe — APPROVED** [review]
LGTM

**Pull request merged**